Source/Details/ASDataController.h

```diff
@@ -136,22 +138,22 @@ extern NSString * const ASCollectionInvalidUpdateException;
  *
  * @discussion This method will be called on main thread.
  */
-- (id)layoutContextWithElements:(ASElementMap *)elements;
+- (ASCollectionLayoutContext *)layoutContextWithElements:(ASElementMap *)elements;
 
 /**
- * @abstract Prepares in advance a new layout with the given context.
+ * @abstract Prepares and returns a new layout for given context.
  *
  * @param context A context that was previously returned by `-layoutContextWithElements:`.
  *
- * @discussion This method is called ahead of time, i.e before the underlying collection/table view is aware of the provided elements.
- * As a result, this method should rely solely on the given context and should not reach out to its collection/table view for information regarding items.
+ * @return The new layout calculated for the given context.
  *
- * @discussion This method will be called on background theads. It must be thread-safe and should not change any internal state of the conforming object.
- * It's recommended to put the resulting layouts of this method into a thread-safe cache that can be looked up later on.
+ * @discussion This method is called ahead of time, i.e before the underlying collection/table view is aware of the provided elements.
+ * As a result, clients must solely rely on the given context and should not reach out to other objects for information not available in the context.
  *
- * @discussion This method must block its calling thread. It can dispatch to other theads to reduce blocking time.
+ * This method will be called on background theads. It must be thread-safe and should not change any internal state of the conforming object.
+ * It must block the calling thread but can dispatch to other theads to reduce total blocking time.
  */
-- (void)prepareLayoutWithContext:(id)context;
+- (ASCollectionLayoutState *)calculateLayoutWithContext:(ASCollectionLayoutContext *)context;
```

Comment:
I played around with this idea and in the end I actually like it. Thanks for the suggestion!